src/request.rs

```diff
@@ -1,220 +1,390 @@
+    ) -> Result<Output, Error>;
+}
 
-    if query.is_empty() {
-        Ok(url.to_string())
-    } else {
-        Ok(format!("{url}?{query}"))
+#[derive(Debug, Clone, Serialize)]
+pub struct IsahcClient;
+
+impl IsahcClient {
+    pub fn new() -> Self {
+        IsahcClient
     }
 }
 
-#[cfg(not(target_arch = "wasm32"))]
-pub(crate) async fn request<
-    Query: Serialize,
-    Body: Serialize,
-    Output: DeserializeOwned + 'static,
->(
-    url: &str,
-    apikey: Option<&str>,
-    method: Method<Query, Body>,
-    expected_status_code: u16,
-) -> Result<Output, Error> {
-    use isahc::http::header;
-    use isahc::http::method::Method as HttpMethod;
-    use isahc::*;
-
-    let builder = Request::builder().header(header::USER_AGENT, qualified_version());
-    let builder = match apikey {
-        Some(apikey) => builder.header(header::AUTHORIZATION, format!("Bearer {apikey}")),
-        None => builder,
-    };
+impl Default for IsahcClient {
+    fn default() -> Self {
+        Self::new()
+    }
+}
 
-    let mut response = match &method {
-        Method::Get { query } => {
-            let url = add_query_parameters(url, query)?;
-
-            builder
-                .method(HttpMethod::GET)
-                .uri(url)
-                .body(())
-                .map_err(|_| crate::errors::Error::InvalidRequest)?
-                .send_async()
-                .await?
-        }
-        Method::Delete { query } => {
-            let url = add_query_parameters(url, query)?;
-
-            builder
-                .method(HttpMethod::DELETE)
-                .uri(url)
-                .body(())
-                .map_err(|_| crate::errors::Error::InvalidRequest)?
-                .send_async()
-                .await?
-        }
-        Method::Post { query, body } => {
-            let url = add_query_parameters(url, query)?;
-
-            builder
-                .method(HttpMethod::POST)
-                .uri(url)
-                .header(header::CONTENT_TYPE, "application/json")
-                .body(to_string(&body).unwrap())
-                .map_err(|_| crate::errors::Error::InvalidRequest)?
-                .send_async()
-                .await?
-        }
-        Method::Patch { query, body } => {
-            let url = add_query_parameters(url, query)?;
-
-            builder
-                .method(HttpMethod::PATCH)
-                .uri(url)
-                .header(header::CONTENT_TYPE, "application/json")
-                .body(to_string(&body).unwrap())
-                .map_err(|_| crate::errors::Error::InvalidRequest)?
-                .send_async()
-                .await?
+#[async_trait(?Send)]
+impl HttpClient for IsahcClient {
+    #[cfg(not(target_arch = "wasm32"))]
+    async fn request<Query, Body, Output>(
+        self,
+        url: &str,
+        apikey: Option<&str>,
+        method: Method<Query, Body>,
+        expected_status_code: u16,
+    ) -> Result<Output, Error>
```

Comment:
Here it would be better to split the two implementations in different block I think;

```rust
// First, the non-wasm impl block where the whole block is gated behind the cfg thingy

#[cfg(not(target_arch = "wasm32"))]
#[async_trait(?Send)]
impl HttpClient for IsahcClient {
async fn request<Query, Body, Output>(
        self,
        url: &str,
       ...) -> Result<Output, Error> { ... }

    async fn stream_request<
        'a,
        Query: Serialize + Send + Sync,
        Body: futures_io::AsyncRead + Send + Sync + 'static,
        Output: DeserializeOwned + 'static,
    >(
        self,
        url: &str,
        ...) -> Result<Output, Error> { ... }
}

// And then the wasm block that doesn’t contains the stream_request method at all

#[cfg(target_arch = "wasm32")]
#[async_trait(?Send)]
impl HttpClient for IsahcClient {
async fn request<Query, Body, Output>(
        self,
        url: &str,
       ...) -> Result<Output, Error> { ... }
}
```

What do you think?

Reply:
It certainly doesn't feel right to have methods with the same name in the same block.
I think it is appropriate to separate them by `flag` and keep the necessary ones, so I've modified it!🙏

https://github.com/meilisearch/meilisearch-rust/pull/459/commits/2a1cc7a8e6b088c15c7c34ac3fe98246c1b96947